core/src/main/scala/org/apache/spark/TaskContextImpl.scala

```diff
@@ -57,57 +68,75 @@ private[spark] class TaskContextImpl(
   // Whether the task has failed.
   @volatile private var failed: Boolean = false
 
+  // Throwable that caused the task to fail
+  private var failure: Throwable = _
+
   // If there was a fetch failure in the task, we store it here, to make sure user-code doesn't
   // hide the exception.  See SPARK-19276
   @volatile private var _fetchFailedException: Option[FetchFailedException] = None
 
+  @GuardedBy("this")
   override def addTaskCompletionListener(listener: TaskCompletionListener): this.type = {
-    onCompleteCallbacks += listener
+    synchronized {
+      if (completed) {
+        listener.onTaskCompletion(this)
+      }
+      // Always add the listener because it is legal to call them multiple times.
```

Comment:
Well, I was rather surprised about this, but the current code path seems to allow this.